### PR TITLE
Add arbitrary trait implementation

### PR DESCRIPTION
**Pull request by kirk-baird**
# What has been done

I've added an implementation of the trait `Arbitrary`. It is hidden behind a feature flag `arbitrary`.

The implementation is pretty much the same as `quickselect` which implements a slightly different `Arbitrary`.

# Additional comments

I'm not sure of the versioning so I left the version as is whether it's worth maybe a patch or minor update.

### CODE REVIEW

File: fixed-hash/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ quickcheck = { version = "0.9.0", optional = true }
 rand = { version = "0.7.2", optional = true, default-features = false }
 rustc-hex = { version = "2.0.1", optional = true, default-features = false }
 static_assertions = "1.0.0"
+arbitrary = { version = "0.3", optional = true }
```
Comment (dvdplm) [inline fixed-hash/Cargo.toml:22]: Curious about the choice of version here, why not use `0.4.2`?

**Comment by kirk-baird** [comment]
[clabot:check]

File: fixed-hash/src/hash.rs [hunk 2]
```diff
@@ -636,6 +637,42 @@ macro_rules! impl_quickcheck_for_fixed_hash {
 	};
 }
 
+// Implementation for disabled rand crate support.
```
Comment (dvdplm) [inline fixed-hash/src/hash.rs:640]:
```suggestion
// When the `arbitrary` features is disabled.
```

File: fixed-hash/src/hash.rs [hunk 3]
```diff
@@ -636,6 +637,42 @@ macro_rules! impl_quickcheck_for_fixed_hash {
...
+// to work around the problems of introducing `arbitrary` crate feature in
+// a user crate.
+#[cfg(not(feature = "arbitrary"))]
+#[macro_export]
+#[doc(hidden)]
+macro_rules! impl_arbitrary_for_fixed_hash {
+	( $name:ident ) => {};
+}
+
+// Implementation for enabled rand crate support.
```
Comment (dvdplm) [inline fixed-hash/src/hash.rs:654]:
```suggestion
// When the `arbitrary` features is disabled.
```

File: fixed-hash/src/hash.rs [hunk 4]
```diff
@@ -636,6 +637,42 @@ macro_rules! impl_quickcheck_for_fixed_hash {
...
+// to work around the problems of introducing `arbitrary` crate feature in
+// a user crate.
+#[cfg(feature = "arbitrary")]
+#[macro_export]
+#[doc(hidden)]
+macro_rules! impl_arbitrary_for_fixed_hash {
+	( $name:ident ) => {
+		impl $crate::arbitrary::Arbitrary for $name {
+			fn arbitrary(u: &mut $crate::arbitrary::Unstructured<'_>) -> $crate::arbitrary::Result<Self> {
+				let mut res = [0u8; $crate::core_::mem::size_of::<Self>()];
```
Comment (dvdplm) [inline fixed-hash/src/hash.rs:668]:
```suggestion
				let mut res = Self::zero();
```

Comment (kirk-baird) [reply fixed-hash/src/hash.rs]: Done :)

File: fixed-hash/Cargo.toml (hunk 1 above)
Comment (kirk-baird) [reply fixed-hash/Cargo.toml:22]: Last I checked, [libfuzzer](https://docs.rs/libfuzzer-sys/0.3.2/libfuzzer_sys/index.html) who is the one of main users of `arbitrary` was pointing to `0.3.0` but looks live they've updated to `0.4.1` now so I'll bump the version to `0.4`

File: fixed-hash/src/hash.rs (hunk 2 above)
Comment (kirk-baird) [reply fixed-hash/src/hash.rs:640]: Done :)

File: fixed-hash/src/hash.rs (hunk 3 above)
Comment (kirk-baird) [reply fixed-hash/src/hash.rs:654]: Done :)

Comment (ordian) [reply fixed-hash/src/hash.rs:654]: you meant enabled :P
to be consistent with other macro impls, I would prefer to keep it, but fix the copy-paste typo `s/rand/arbitrary`

File: ethereum-types/Cargo.toml [hunk 5]
```diff
@@ -22,3 +22,4 @@ serde_json = "1.0.41"
 default = ["std", "serialize"]
 std = ["uint-crate/std", "fixed-hash/std", "ethbloom/std", "primitive-types/std"]
 serialize = ["std", "impl-serde", "primitive-types/serde", "ethbloom/serialize"]
+arbitrary = ["std", "serialize", "fixed-hash/arbitrary"]
```
Comment (ordian) [inline ethereum-types/Cargo.toml:25]: why is `serialize` required?

Comment (ordian) [reply ethereum-types/Cargo.toml:25]: would you mind adding `arbitrary` support for `uint` as well?

Comment (kirk-baird) [reply ethereum-types/Cargo.toml:25]: Good point neither of `std` or `serialize` are actually required but were useful to me so I'll remove them. It's easy for the calling library to have `["default", "aribtrary"]`

**Comment by kirk-baird** [comment]
> Thanks!
> 
> Can you please document the feature in the README?

My pleasure! It's updated.

File: ethereum-types/Cargo.toml (hunk 5 above)
Comment (kirk-baird) [reply ethereum-types/Cargo.toml]: > why is `serialize` required?

Yep can do

Comment (kirk-baird) [reply ethereum-types/Cargo.toml]: > would you mind adding `arbitrary` support for `uint` as well?

Yep can do

**Comment by ordian** [comment]
`fixed-hash` is used in exposed in several crates here, like `primitive-types`, `ethbloom`, do you think `arbitrary` feature should be propagated to all these crates?

**Comment by kirk-baird** [comment]
> `fixed-hash` is exposed in several crates here, like `primitive-types`, `ethbloom`, do you think `arbitrary` feature should be propagated (re-exported) to all these crates?

I think `primitive-types` is a good idea in case anyone is attempting to fuzz a struct which uses one. `ethbloom` less so though happy to include it for completeness.

File: uint/src/uint.rs [hunk 6]
```diff
@@ -1632,3 +1633,26 @@ macro_rules! impl_quickcheck_arbitrary_for_uint {
...
+#[cfg(feature = "arbitrary")]
+#[macro_export]
+#[doc(hidden)]
+macro_rules! impl_arbitrary_for_uint {
+	($uint: ty, $n_bytes: tt) => {
+		impl $crate::arbitrary::Arbitrary for $uint {
+			fn arbitrary(u: &mut $crate::arbitrary::Unstructured<'_>) -> $crate::arbitrary::Result<Self> {
+				let mut res = [0u8; $n_bytes];
+				u.fill_buffer(&mut res)?;
+				Ok(Self::from(res))
```
Comment (ordian) [inline uint/src/uint.rs:1647]:
```suggestion
				let mut res = Self::zero();
				u.fill_buffer(&mut res.0)?;
				Ok(res)
```

File: ethereum-types/Cargo.toml [hunk 7]
```diff
@@ -22,3 +22,4 @@ serde_json = "1.0.41"
 default = ["std", "serialize"]
 std = ["uint-crate/std", "fixed-hash/std", "ethbloom/std", "primitive-types/std"]
 serialize = ["std", "impl-serde", "primitive-types/serde", "ethbloom/serialize"]
+arbitrary = ["fixed-hash/arbitrary", "uint-crate/arbitrary"]
```
Comment (ordian) [inline ethereum-types/Cargo.toml:25]:
```suggestion
arbitrary = ["ethbloom/arbitrary", "fixed-hash/arbitrary", "uint-crate/arbitrary"]
```

File: uint/src/uint.rs (hunk 6 above)
Comment (kirk-baird) [reply uint/src/uint.rs:1647]: Unfortunately the `unstructured.fill_buffer(&bytes)` will only take `u8` and so this won't work as the word sizes are `u64`.

Comment (ordian) [reply uint/src/uint.rs:1647]: Ah, right, good point.

File: uint/README.md [hunk 8]
```diff
@@ -69,3 +69,5 @@ see fuzz [README.md](fuzz/README.md)
 	- Enabled by default.
 - `quickcheck`: Enable quickcheck-style property testing
 	- Use with `cargo test --release --features=quickcheck`.
+- `arbitrary`: Allow for creation of an `uint` object from random unstructured input.
```
Comment (dvdplm) [inline uint/README.md:72]:
```suggestion
- `arbitrary`: Allow for creation of an `uint` object from random unstructured input for use with fuzzers that use the `arbitrary` crate.
```

Also, please document the feature in all the readmes in crates where you add it.

File: fixed-hash/src/hash.rs [hunk 9]
```diff
@@ -651,7 +651,7 @@ macro_rules! impl_arbitrary_for_fixed_hash {
 	( $name:ident ) => {};
 }
 
-// Implementation for enabled rand crate support.
+// Implementation for enabled `arbitrary` crate support.
```
Comment (dvdplm) [inline fixed-hash/src/hash.rs:654]: I know this is consistent with the rest of the comments but it's grammatically incorrect and hard to parse. Please consider using the previous suggestion (or rewrite it to your pleasing). It's fine if you correct only the additions in this PR.

**Comment by dvdplm** [comment]
> I'm not sure of the versioning so I left the version as is whether it's worth maybe a patch or minor update.

This is a patch update I'd say. What we usually do is add unpublished changes to the `CHANGELOG` under the "Unreleased" header. When the time comes for publication we'll bump the version and fix up the change log.

Speaking of releases: do you have an immediate need for this?

File: fixed-hash/src/hash.rs (hunk 9 above)
Comment (kirk-baird) [reply fixed-hash/src/hash.rs]: No worries, I've updated the two lines for arbitrary so that they are grammatically correct.

**Comment by kirk-baird** [comment]
> This is a patch update I'd say. What we usually do is add unpublished changes to the `CHANGELOG` under the "Unreleased" header. When the time comes for publication we'll bump the version and fix up the change log.

Do I need to make any changes to the `CHANGELOG` or will you guys handle that?
 
> Speaking of releases: do you have an immediate need for this?

It's not mission critical, I'm currently using a patch to point to my branch.

File: fixed-hash/src/hash.rs [hunk 10]
```diff
@@ -637,7 +637,7 @@ macro_rules! impl_quickcheck_for_fixed_hash {
 	};
 }
 
-// Implementation for disabled `arbitrary` crate support.
+// When the `arbitrary` features is disabled.
```
Comment (dvdplm) [inline fixed-hash/src/hash.rs:640]:
```suggestion
// When the `arbitrary` feature is disabled.
```

File: fixed-hash/src/hash.rs [hunk 11]
```diff
@@ -651,7 +651,7 @@ macro_rules! impl_arbitrary_for_fixed_hash {
 	( $name:ident ) => {};
 }
 
-// Implementation for enabled `arbitrary` crate support.
+// When the `arbitrary` features is enabled.
```
Comment (dvdplm) [inline fixed-hash/src/hash.rs:654]:
```suggestion
// When the `arbitrary` feature is enabled.
```

**Comment by ordian** [comment]
> Do I need to make any changes to the CHANGELOG or will you guys handle that?

We can handle that, no worries.
